src/main/java/com/khomishchak/ws/services/integration/whitebit/WhiteBitExchangerConnectorService.java

```diff
@@ -1,26 +1,107 @@
+        this.responseMapper = responseMapper;
+        this.balanceRepository = balanceRepository;
+        this.depositWithdrawalTransactionsHistoryRepository = depositWithdrawalTransactionsHistoryRepository;
     }
 
     @Override
     public Balance getMainBalance(long userId) {
-        return whiteBitService.getAccountBalance(userId);
+        return whiteBitService.getAccountBalance(userId)
+                .map(balanceResp -> handleSuccessfulBalanceResp(userId, balanceResp))
+                .block();
     }
 
     @Override
     public ExchangerDepositWithdrawalTransactions getDepositWithdrawalHistory(long userId) {
-        return whiteBitService.getDepositWithdrawalHistory(userId);
+        return whiteBitService.getDepositWithdrawalHistory(userId)
+                .map(transactionResp -> handleSuccessfulDepositWithdrawalTransactionsResp(userId, transactionResp))
+                .block();
+    }
+
+    private Balance handleSuccessfulBalanceResp(long userId, WhiteBitBalanceResp response) {
+        List<Currency> availableCurrencies = responseMapper.mapToCurrencies(response);
+        return getPopulatedWithCurrenciesBalance(userId, availableCurrencies);
+    }
+
+    private Balance getPopulatedWithCurrenciesBalance(long userId, List<Currency> availableCurrencies) {
+        Balance balance = getBalanceByUserIdOrThrowException(userId);
+        balance.setCurrencies(availableCurrencies);
+        balance.setLastTimeWasUpdated(LocalDateTime.now());
+        return balance;
+    }
+
+    private ExchangerDepositWithdrawalTransactions handleSuccessfulDepositWithdrawalTransactionsResp(long userId,
+                                                                    WhiteBitDepositWithdrawalHistoryResp response) {
+        List<DepositWithdrawalTransaction> transactions = responseMapper.mapWithdrawalDepositHistoryToTransactions(response);
+        Balance balance = getBalanceByUserIdOrThrowException(userId);
+        return getExchangerDepositWithdrawalTransactions(balance, transactions);
+    }
+
+    private ExchangerDepositWithdrawalTransactions getExchangerDepositWithdrawalTransactions(Balance balance,
+                                                                    List<DepositWithdrawalTransaction> transactions) {
+        ExchangerDepositWithdrawalTransactions exchangerTransactions = getExchangerDepositWithdrawalTransactions(balance);
+        assigneeTransactionsToExchangerTransactionsEntity(exchangerTransactions, transactions);
+        return exchangerTransactions;
+    }
+
+    private ExchangerDepositWithdrawalTransactions getExchangerDepositWithdrawalTransactions (Balance balance) {
+        return depositWithdrawalTransactionsHistoryRepository.findByBalance_Id(balance.getId())
+                .orElseGet(() -> getNewExchangerDepositWithdrawalTransactions(balance));
+    }
+
+    private ExchangerDepositWithdrawalTransactions getNewExchangerDepositWithdrawalTransactions(Balance balance) {
```

Comment:
Isn't better to move it to entity layer ?

Reply:
I prefer to not to have any business related logic on entity layer